File: 443.string-compression.py
#
# @lc app=leetcode id=443 lang=python3
#
# [443] String Compression
#

# @lc code=start
class Solution:
    def compress(self, chars: List[str]) -> int:
        if len(chars)==1:
            return 1
        currChar=chars[0]
        breakpoints=[[0, currChar]] # [index, char] pair
        for i, char in enumerate(chars):
            if i==0:
                continue
            if char != currChar:
                breakpoints.append([i, char])
                currChar = char
        
        s=""
        for i, pair in enumerate(breakpoints):
            index, char = pair
            nextIndex = len(chars) if i==len(breakpoints)-1 else breakpoints[i+1][0]
            length = nextIndex-index
            s+=char
            if length>1:
                s+=str(length)
        
        for i in range(len(s)):
            chars[i]=s[i]

        return len(s)
            

        
# @lc code=end

File: 443.string-compression.go
/*
 * @lc app=leetcode id=443 lang=golang
 *
 * [443] String Compression
 */

// @lc code=start
package main

import (
	"strconv"
)

func compress(chars []byte) int {
	read_idx := 0
	write_idx := 0
	n := len(chars)
	for read_idx < n {
		char := chars[read_idx]
		// get how many char
		times := 0
		for read_idx < n && chars[read_idx] == char {
			read_idx++
			times++
		}
		chars[write_idx] = char
		write_idx++
		if times > 1 {
			times_str := strconv.Itoa(times)
			for j := 0; j < len(times_str); j++ {
				chars[write_idx] = times_str[j]
				write_idx++
			}
		}
	}
	return write_idx
}

// @lc code=end
